### PR TITLE
Also try mangling sdist urls to match pep625

### PR DESCRIPTION
**Pull request by bollwyvl**
- [x] Pydantic model updated or no update needed

This adds _yet another_ URL to try, namely the basically-now-ubiquitous [PEP625](https://peps.python.org/pep-0625/) form:

```
old    package-name-0.0.0.tar.gz
              V
new    package_name-0.0.1.tar.gz
```

### CODE REVIEW

**Review by beckermr — CHANGES_REQUESTED** [review]
Can we add some tests? I know the tests are thing already. :/

**Comment by bollwyvl** [comment]
> add some tests? I know the tests are thing already. :/

how do we feel about moving a bunch of the `black`-violating tests to `.txt`-based fixtures?

**Comment by beckermr** [comment]
What do you mean by that? W/e it is, a new PR please.

File: conda_forge_tick/url_transforms.py [hunk 1]
```diff
@@ -48,23 +50,29 @@ def _v_munger(url):
...
+    has_version = re.search(r"\{\{\s*version", bn)
 
+    # try the original url first, as a fallback (probably can't be removed?)
     yield url
 
+    if not (is_sdist and is_pypi and has_version):
+        return
+
+    # try static PEP625 name with PEP345 distribution name (_ not -)
+    yield os.path.join(dn, '%s-{{ version }}.tar.gz' % dist_bn.replace("-", "_"))
```
Comment (jaimergp) [inline conda_forge_tick/url_transforms.py:72]: I think the PEP normalization also transforms periods to underscores. Check with ruamel.yaml to confirm but in that case we will need one more yield here.

File: conda_forge_tick/url_transforms.py [hunk 2]
```diff
@@ -141,7 +149,9 @@ def gen_transformed_urls(url):
     url : str
         The URL to transform.
     """
-    yield from _gen_new_urls(
+    yielded = []
```
Comment (jaimergp) [inline conda_forge_tick/url_transforms.py:152]: I guess this could be a set.

File: conda_forge_tick/url_transforms.py [hunk 3]
```diff
@@ -154,4 +164,7 @@ def gen_transformed_urls(url):
             _pypi_name_munger,
             _github_munger,
         ],
-    )
+    ):
+        if new_url not in yielded:
+            yield new_url
+            yielded.append(new_url)
```
Comment (jaimergp) [inline conda_forge_tick/url_transforms.py:170]: And in that case this would be `.add`

**Comment by beckermr** [comment]
I honestly don't get how this PR passes or works. Here is the existing tests for url transforms  w/pypi (which is way more complete than I recall!):

```
def test_url_transform_pypi():
    urls = set(list(gen_transformed_urls("https://pypi.io/{{ name }}/{{ name }}-barf")))
    assert urls == {
        "https://files.pythonhosted.org/{{ name }}/{{ name }}-barf",
        "https://files.pythonhosted.org/{{ name }}/{{ name.replace('-', '_') }}-barf",
        "https://files.pythonhosted.org/{{ name }}/{{ name.replace('_', '-') }}-barf",
        "https://pypi.io/{{ name }}/{{ name }}-barf",
        "https://pypi.io/{{ name }}/{{ name.replace('-', '_') }}-barf",
        "https://pypi.io/{{ name }}/{{ name.replace('_', '-') }}-barf",
    }

    urls = set(
        list(
            gen_transformed_urls(
                "https://pypi.io/{{ name }}/{{ name.replace('_', '-') }}-barf",
            ),
        ),
    )
    assert urls == {
        "https://files.pythonhosted.org/{{ name }}/{{ name }}-barf",
        "https://files.pythonhosted.org/{{ name }}/{{ name.replace('_', '-') }}-barf",
        "https://pypi.io/{{ name }}/{{ name }}-barf",
        "https://pypi.io/{{ name }}/{{ name.replace('_', '-') }}-barf",
    }

    urls = set(
        list(
            gen_transformed_urls(
                "https://pypi.io/{{ name }}/{{ name.replace('_','-') }}-barf",
            ),
        ),
    )
    assert urls == {
        "https://files.pythonhosted.org/{{ name }}/{{ name }}-barf",
        "https://files.pythonhosted.org/{{ name }}/{{ name.replace('_','-') }}-barf",
        "https://pypi.io/{{ name }}/{{ name }}-barf",
        "https://pypi.io/{{ name }}/{{ name.replace('_','-') }}-barf",
    }

    urls = set(
        list(
            gen_transformed_urls(
                "https://pypi.io/{{ name }}/{{ name|replace('_','-') }}-barf",
            ),
        ),
    )
    assert urls == {
        "https://files.pythonhosted.org/{{ name }}/{{ name }}-barf",
        "https://files.pythonhosted.org/{{ name }}/{{ name|replace('_','-') }}-barf",
        "https://pypi.io/{{ name }}/{{ name }}-barf",
        "https://pypi.io/{{ name }}/{{ name|replace('_','-') }}-barf",
    }

    urls = set(
        list(
            gen_transformed_urls(
                'https://pypi.io/{{ name }}/{{ name.replace("_", "-") }}-barf',
            ),
        ),
    )
    assert urls == {
        "https://files.pythonhosted.org/{{ name }}/{{ name }}-barf",
        'https://files.pythonhosted.org/{{ name }}/{{ name.replace("_", "-") }}-barf',
        "https://pypi.io/{{ name }}/{{ name }}-barf",
        'https://pypi.io/{{ name }}/{{ name.replace("_", "-") }}-barf',
    }

```

We already test at least part of the `-` to `_` transform, right?

**Comment by beckermr** [comment]
Oh you are munging hard-coded names. duh. Needs tests then for sure.

**Comment by bollwyvl** [comment]
> fixtures

> url transforms w/ pypi (which is way more complete than I recall!):

Made #2797. Might do a little work there before coming back to here... and indeed, this (draft) PR needs some better stress tests, especially of bad/special/magic cases that categorically _did_ change over time (PEP420 with `.`, `poetry` and `flit` epochs, etc. ).

**Comment by beckermr** [comment]
We have pre-commit so it should be easy to solve linter errors.

**Comment by beckermr** [comment]
pre-commit.ci autofix

**Comment by bollwyvl** [comment]
I am struggling to find examples of any in-the-wild examples that normalizes `.` to `_`, or apply case normalization.

**Comment by beckermr** [comment]
That's fine. As long as something made up is in the tests, it will be fine I think.

File: conda_forge_tick/url_transforms.py (hunk 3 above)
Comment (bollwyvl) [reply conda_forge_tick/url_transforms.py]: edcfef82

File: conda_forge_tick/url_transforms.py [hunk 4]
```diff
@@ -48,23 +50,38 @@ def _v_munger(url):
...
         yield url.replace(old_d, new_d, 1)
 
 
 def _pypi_name_munger(url):
     bn = os.path.basename(url)
-    if (
-        url.startswith("https://pypi.io")
-        or url.startswith("https://files.pythonhosted.org")
-    ) and ("{{ version }}" in bn and "{{ name" not in bn):
-        yield os.path.join(os.path.dirname(url), "{{ name }}-{{ version }}.tar.gz")
```
Comment (beckermr) [inline conda_forge_tick/url_transforms.py:64]: It looks like you removed this variant that we try. We should yield it too.

File: conda_forge_tick/url_transforms.py (hunk 1 above)
Comment (bollwyvl) [reply conda_forge_tick/url_transforms.py]: [4face02](https://github.com/regro/cf-scripts/pull/2794/commits/4face02266bc5b2d2035c48ee5707bb08956e037)

**Review by beckermr — CHANGES_REQUESTED** [review]

File: conda_forge_tick/url_transforms.py (hunk 4 above)
Comment (bollwyvl) [reply conda_forge_tick/url_transforms.py:64]: Hm... it doesn't check to see if that variable is even set... seems like _adding_ variables when not asked could potentially make things _worse_...

Comment (beckermr) [reply conda_forge_tick/url_transforms.py:64]: If the url doesn't work it tries the next one, so it doesn't matter. Also there is a test if name is in the url in the first place. That means it is likely there in the recipe. 

In general though, please do not remove features. I've seen this more than once in PRs others have made to the bot and I don't really understand it. 

Everything is here for a reason, even if we do not recall why. Unless we know what happened, we should leave things in place. 

Let's add this back and then we can merge.

Comment (bollwyvl) [reply conda_forge_tick/url_transforms.py:64]: 2bea2c2f

**Review by beckermr — APPROVED** [review]

**Comment by beckermr** [comment]
Thank you!

**Pull request merged**